runtime/upgrade.go

```diff
@@ -0,0 +1,129 @@
+package runtime
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/alibaba/sealer/client"
+	"github.com/alibaba/sealer/logger"
+	"github.com/alibaba/sealer/utils"
+)
+
+func (d *Default) upgrade() error {
+	err := d.UpgradeMasters0()
+	if err != nil {
+		return err
+	}
+
+	if len(d.Masters) > 1 {
+		err = d.UpgradeOtherMasters()
+	}
+
+	if len(d.Nodes) >= 1 {
+		err = d.UpgradeNodes()
+	}
+
+	return err
+}
+
+func (d *Default) UpgradeMasters0() error {
+	logger.Info("Upgrade Masters0, %s", d.Masters[:1])
+	return d.upgradeNodes(d.Masters[:1], true)
+}
+
+func (d *Default) UpgradeOtherMasters() error {
+	logger.Info("UpgradeMasters")
+	return d.upgradeNodes(d.Masters[1:], true)
+}
+
+func (d *Default) UpgradeNodes() error {
+	logger.Info("UpgradeNodes")
+	return d.upgradeNodes(d.Nodes, false)
+}
+
+func (d *Default) upgradeNodes(hosts []string, isMaster bool) error {
+	wg := NewPool(2)
+	var err error
+	c, err := client.NewClientSet()
+	if err != nil {
+		logger.Info("current cluster not found, upgrade nothing %v", err)
+		return err
+	}
+	var errList []error
+	for _, host := range hosts {
+		wg.Add(1)
+		go func(h string) {
+			defer wg.Done()
+			node := d.GetRemoteHostName(h)
+			ip := utils.GetHostIP(h)
+			// drain worker node is too danger for prod use; do not drain nodes if worker nodes~
+			if isMaster {
+				logger.Info("[%s] first: to drain master node %s", ip, node)
+				cmdDrain := fmt.Sprintf(`kubectl drain %s --ignore-daemonsets --delete-local-data`, node)
+				err := d.SSH.CmdAsync(d.Masters[0], cmdDrain)
+				if err != nil {
+					logger.Error("kubectl drain %s  err: %v", node, err)
+					errList = append(errList, err)
+				}
+			} else {
+				logger.Info("first: to print upgrade node %s", node)
+			}
+
+			logger.Info("first: to print upgrade node %s", node)
+
+			// second to exec kubeadm upgrade node
+			logger.Info("[%s] second: to exec kubeadm upgrade node on %s", ip, node)
+			var cmdUpgrade string
+			if ip == d.Masters[0] {
+				cmdUpgrade = fmt.Sprintf("kubeadm upgrade apply --certificate-renewal=false  --yes %s", d.Version)
+				err = d.SSH.CmdAsync(ip, cmdUpgrade)
+				if err != nil {
+					// master1 upgrade failed exit.
+					logger.Error("kubeadm upgrade err: ", err)
+					errList = append(errList, err)
+					os.Exit(1)
```

Comment:
please do not exit the programming execution in implementation ever. You can make the caller to judge the returned error. 

As detailed function implementation, it usually has no rights to decide program exit except some particular and harsh cases.

In addition, you directly exit, all context has abandoned. For example,  the previous line of code `errList = append(errList, err)` is redudant then.

Always be careful with exit in detailed implementation code.